lib/ads/client_side_ad_manager.js

```diff
@@ -406,6 +406,8 @@ shaka.ads.ClientSideAdManager = class {
         'Should have an ads manager at this point!');
 
     const imaAd = e.getAd();
+    goog.asserts.assert(imaAd,
+        e.type + ' event from IMA SDK should have an associated ad!');
```

Comment:
I've seen this scenario occur with IMA. Is an assert the best option here? Won't that throw an error and halt playback? Most players will continue on with content if there is an error or ad-less ad break. Would wrapping this in a `null` check, like `onAdComplete_` below, be a better fix?

Reply:
In production builds, assertions are removed.  In debug builds, assertions log, but don't throw.  In tests, assertions log and fail the current test, but still don't throw.

Reply:
Theodore, I don't see what this is meant to accomplish.  If I read the issue correctly, there was an expectation thrown on `isLinear()`, right?  (this.ad_?)

Reply:
Correct. Our code assumes that `e.getAd()` returns a `google.ima.Ad` object.
However, in some situations it appears to return null instead. If it is null, then a `TypeError` goes into the console, which doesn't really communicate what has actually gone wrong.

According to their documentation (https://developers.google.com/interactive-media-ads/docs/sdks/html5/client-side/reference/js/google.ima.AdEvent#getAd) the `getAd` method returns a nullable ad, so we should probably also change our extern for IMA too. I'll add that to this PR.

Anyway, I don't know why IMA SDK would be returning null for these events. The general scenarios, as far as I can tell, are:

1.  It's the result of the IMA SDK being misused or mis-configured by the user in some way. I feel an assert would probably be the right solution in this case, to communicate to the developer that there is something going wrong (instead of having the ads silently fail or something).

1. It's possible there's some bug in the IMA SDK that's causing this. In that case I guess we will have to come up with a workaround, at least until it's fixed.

1. Maybe we're interpreting the IMA SDK wrong, and the `STARTED` event is actually meant to fire with no ad in some situation or another. Their tutorial [doesn't make it sound that way](https://developers.google.com/interactive-media-ads/docs/sdks/html5/client-side#triggering-play-on-non-linear-ads), though.

I decided to go with an assert because it was the lowest-impact change. It's not going to break anyone's build, if it turns out to be the wrong change later.

Reply:
In my experience, the IMA SDK does not always behave sensically. You can get valid ad responses that have no playable ads for a given break, but you get events anyway.

Also, because the SDKs URL is _not_ versioned, they release breaking changes and bugs that aren't found until production sites encounter runtime errors (I've been pulled into many emergency meetings because of this). 

I still believe it would be better to add a null check like the ad end function.

Reply:
Fair enough. If this is such a common scenario, maybe handling it recoverably would be best.
I'll just have an `alwaysWarn` log and ignore the event.

Just in case, I'll also change a few other asserts that assume we have an ad when other events fire. If we are expecting them to fire `STARTED` events with no ad, I'm worried they might also fire `SKIP_STATE_CHANGED` events and such after that non-ad.